File: Sources/AndroidBackend/CustomSlider.swift
import AndroidKit
import SwiftJava

@JavaClass("dev.swiftcrossui.androidbackend.CustomSlider")
class CustomSlider: JavaObject {
    @JavaMethod
    @_nonoverride convenience init(
        _ activity: AndroidKit.Activity!,
        environment: JNIEnvironment? = nil
    )

    @JavaMethod
    func setAction(_ action: SwiftAction?)

    @JavaMethod
    func setBounds(min: Float, max: Float, places: Int32)

    // Inherited from Slider
    @JavaMethod
    func getValue() -> Float

    @JavaMethod
    func setValue(_ value: Float)

    // Inherited from BaseSlider
    @JavaMethod
    func setEnabled(_ enabled: Bool)
}

File: Sources/AndroidBackend/Kotlin/CustomSlider.kt
package dev.swiftcrossui.androidbackend

import android.app.Activity
import com.google.android.material.slider.Slider

class CustomSlider(activity: Activity) : Slider(activity) {
    var action: SwiftAction? = null

    private var places = 7

    init {
        addOnChangeListener { _, _, fromUser ->
            if (fromUser) {
                action?.call()
            }
        }

        isTickVisible = false

        setLabelFormatter { String.format("%.${places}f", it.toDouble()) }
    }

    fun setBounds(min: Float, max: Float, places: Int) {
        this.places = places
        valueFrom = min
        valueTo = max
    }
}
